File: 1895-largest-magic-square/1895-largest-magic-square.cpp
class Solution {
private:
    vector<vector<int>> rowSum;
    vector<vector<int>> colSum;
    bool checkSquare(vector<vector<int>>& grid, int row, int col, int side) {
        if (row + side - 1 >= rowSum.size() ||
            col + side - 1 >= colSum.size()) {
            return false;
        }
        int compSum =
            rowSum[row][col + side - 1] - rowSum[row][col] + grid[row][col];
        for (int i = row + 1; i < row + side; i++) {
            int sum = rowSum[i][col + side - 1] - rowSum[i][col] + grid[i][col];
            if (sum != compSum) {
                return false;
            }
        }
        for (int j = col; j < col + side; j++) {
            int sum = colSum[j][row + side - 1] - colSum[j][row] + grid[row][j];
            if (sum != compSum) {
                return false;
            }
        }
        int sumDiag1 = 0;
        int sumDiag2 = 0;
        for (int i = row, j = col; i < row + side; i++, j++) {
            sumDiag1 += grid[i][j];
        }
        for (int i = row, j = col + side - 1; i < row + side; i++, j--) {
            sumDiag2 += grid[i][j];
        }
        if (sumDiag1 != compSum || sumDiag2 != compSum) {
            return false;
        }
        return true;
    }

public:
    int largestMagicSquare(vector<vector<int>>& grid) {
        int res = 1;
        int n = grid.size();
        int m = grid[0].size();
        int maxSide = min(n, m);
        rowSum = vector<vector<int>>(n, vector<int>(m, 0));
        colSum = vector<vector<int>>(m, vector<int>(n, 0));
        for (int i = 0; i < n; i++) {
            int sum = 0;
            for (int j = 0; j < m; j++) {
                sum += grid[i][j];
                rowSum[i][j] = sum;
            }
        }
        for (int j = 0; j < m; j++) {
            int sum = 0;
            for (int i = 0; i < n; i++) {
                sum += grid[i][j];
                colSum[j][i] = sum;
            }
        }
        for (int side = 2; side <= maxSide; side++) {
            bool flag = false;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (checkSquare(grid, i, j, side)) {
                        res = side;
                        flag = true;
                        break;
                    }
                }
                if(flag)
                {
                    break;
                }
            }
        }
        return res;
    }
};

File: 1895-largest-magic-square/1895-largest-magic-square.py
class Solution:
    def largestMagicSquare(self, grid: list[list[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        rowSum = [[0] * m for _ in range(n)]
        colSum = [[0] * n for _ in range(m)]
        for i in range(n):
            current_sum = 0
            for j in range(m):
                current_sum += grid[i][j]
                rowSum[i][j] = current_sum    
        for j in range(m):
            current_sum = 0
            for i in range(n):
                current_sum += grid[i][j]
                colSum[j][i] = current_sum
        def checkSquare(row, col, side):
            compSum = rowSum[row][col + side - 1] - rowSum[row][col] + grid[row][col]
            for i in range(row + 1, row + side):
                current_row_sum = rowSum[i][col + side - 1] - rowSum[i][col] + grid[i][col]
                if current_row_sum != compSum:
                    return False
            for j in range(col, col + side):
                current_col_sum = colSum[j][row + side - 1] - colSum[j][row] + grid[row][j]
                if current_col_sum != compSum:
                    return False
            sumDiag1 = 0
            for k in range(side):
                sumDiag1 += grid[row + k][col + k]
            if sumDiag1 != compSum:
                return False
            sumDiag2 = 0
            for k in range(side):
                sumDiag2 += grid[row + k][col + side - 1 - k]
            if sumDiag2 != compSum:
                return False     
            return True
        res = 1
        maxSide = min(n, m)
        for side in range(2, maxSide + 1):
            flag = False
            for i in range(n - side + 1):
                for j in range(m - side + 1):
                    if checkSquare(i, j, side):
                        res = side
                        flag = True
                        break
                if flag:
                    break         
        return res